### PR TITLE
Fix scalar arithemetic and add test cases

### PR DESCRIPTION
**Pull request by Erik-Lundell**
Add UnsquezeScalarPlaceholders pass to make scalars rank 1 Add MatchShapesPass to guarantee same rank for all inputs for ops that require it.
Additional fixes to make Scalar tests pass

Map which cases work and which don't.

Signed-off-by: Erik Lundell <erik.lundell@arm.com>
Change-Id: I4ea5e189e26cf7aff391ec153d525b2fb61aa16f

Fix shape issues

Change-Id: I0b8588cd5f8b284c25e806bb83bc788067d5b649

### CODE REVIEW

File: backends/arm/_passes/match_shapes_pass.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                if not isinstance(arg, Node):
+                    continue
+                shape = get_first_fake_tensor(arg).shape
+                rank = len(shape)
+                if rank == max_rank:
+                    continue
+
+                # If the argument is call_function, match shape by inserting view node.
+                if arg.op == "call_function":
+                    self._match_op_shape(graph_module, node, arg, rank, max_rank, shape)
```
Comment (digantdesai) [inline backends/arm/_passes/match_shapes_pass.py:105]: can't add/sub handle broadcasting in TOSA? This view will become a copy and I am not sure if we can do faster broadcasting than this. scalr to 1D makes sense I am talking about matching shape. Ok with following up on this later if you want to change that is.

File: backends/arm/_passes/match_shapes_pass.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    """
+    For ops in 'targeted_ops', make sure that the inputs share the same rank.
+    In particular, unsqueeze inputs of rank 1.
+    """
+
+    def __init__(self, exported_program):
+        super().__init__()
+        self.exported_program = exported_program
+
+    targeted_ops = [
```
Comment (digantdesai) [inline backends/arm/_passes/match_shapes_pass.py:31]: no other ops can take scalar?

**Review by digantdesai — APPROVED** [review]

**Comment by digantdesai** [comment]
Internal errors

executorch/backends/arm/_passes/match_shapes_pass.py:80:4 Inconsistent override [15]: `executorch.backends.arm._passes.match_shapes_pass.MatchShapesPass.call` overrides method defined in `executorch.exir.pass_base._ExportPassBase` inconsistently. Returned type `torch.fx.graph_module.GraphModule` is not a subtype of the overridden return `torch.fx.passes.infra.pass_base.PassResult`.

executorch/backends/arm/_passes/match_shapes_pass.py:113:8 Incompatible return type [7]: Expected `GraphModule` but got `PassResult`.

File: backends/arm/_passes/match_shapes_pass.py (hunk 1 above)
Comment (Erik-Lundell) [reply backends/arm/_passes/match_shapes_pass.py:105]: TOSA handles broadcasting but still requires that the input tensors share the same rank (see https://www.mlplatform.org/tosa/tosa_spec.html#_add). So the pass is actually badly named, should be match_rank. will update.

File: backends/arm/_passes/match_shapes_pass.py (hunk 2 above)
Comment (Erik-Lundell) [reply backends/arm/_passes/match_shapes_pass.py:31]: The pass might be relevant for other ops but the reasoning is to only target what we know we need and what we test. I.e. to limit the scope of the change.

**Pull request closed without merging**